go/bazel.go

```diff
@@ -0,0 +1,158 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package bazel_go
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"syscall"
+)
+
+var _ = fmt.Printf
+
+var bazelVersion = "bazel_version was not set. This is an error."
+
+type TestingBazel interface {
+	ScratchDir(name string) error
+	ScratchFile(name, content string) error
+	ScratchFileWithMode(name, content string, mode os.FileMode) error
+	RunBazel(args []string) (exit_code int, stdout string, stderr string)
+	Rlocation(runfile string) string
+	IsWindows() bool
+}
+
+type testingBazel struct {
+	// Path to the Bazel binary.
+	bazel string
+
+	// Path to the testing temp directory.
+	tmpDir string
+}
+
+// New instance of a TestingBazel will be created and the program will cd into
+// the appropriate working dir for that bazel client.
+func New() (TestingBazel, error) {
+	// Make a temporary dir in the OS specific tmpdir.
+	dir, err := ioutil.TempDir("", "testing")
+	if err != nil {
+		return nil, errors.New("Unable to make tempdir for bazel")
+	}
+
+	err = os.Chdir(dir)
+	if err != nil {
+		return nil, errors.New("Unable to make tempdir for bazel")
+	}
+
+	b := &testingBazel{
+		tmpDir: dir,
+	}
+	if !b.setAndUnpackBazel(fmt.Sprintf("build_bazel_bazel_%s/bazel", strings.Replace(bazelVersion, ".", "_", -1))) {
+		return nil, errors.New("Unable to find Bazel")
+	}
+
+	return b, nil
+}
+
+func (b *testingBazel) setAndUnpackBazel(bazel string) bool {
+	bazel = b.Rlocation(bazel)
+	if _, err := os.Stat(bazel); !os.IsNotExist(err) {
+		b.bazel = bazel
+		// Unpack Bazel by running a command on it.
+		b.RunBazel([]string{"help"})
+		return true
+	}
+	return false
+}
+
+// ScratchFile creates a new file at the path you provided.
+func (b *testingBazel) ScratchDir(p string) error {
+	return os.MkdirAll(p, 0755)
+}
+
+// ScratchFile creates a new file at the path you provided. If the file is in a
+// directory that may not exist, use ScratchDir to make the directory. To set a
+// mode on the file use ScratchFileWithMode.
+func (b *testingBazel) ScratchFile(p, content string) error {
+	return b.ScratchFileWithMode(p, content, 0644)
+}
+
+// ScratchFile creates a new file at the path you provided with a specified
+// mode (useful for setting executable bit). If the file is in a directory that
+// may not exist, use ScratchDir to make the directory.
+func (b *testingBazel) ScratchFileWithMode(p, content string, mode os.FileMode) error {
+	if err := ioutil.WriteFile(p, []byte(content), mode); err != nil {
+		return err
+	}
+	return nil
+}
+
+// GetBazel returns the path to the Bazel binary that is available.
+func (b *testingBazel) GetBazel() string {
+	return b.bazel
+}
+
+func (b *testingBazel) RunBazel(args []string) (int, string, string) {
+	find := exec.Command("ls", "-laR")
```

Comment:
This looks like debug code. Can it be removed?

Reply:
Done